### PR TITLE
LibArchive implementation

### PR DESCRIPTION
**Pull request by juliospain**
Libarchive file system implementation.

PS: first pull request I have ever done, so probably some things are wrongly done...?

### CODE REVIEW

**Comment by martindurant** [comment]
Thank you for contributing!

A couple of quick notes first:
- there ought to be tests, which will involve creating the file test_libarchive.py in fsspec/implementations/tests/ , and should hit as much of your code as possible. You will need to add libarchive itself to tox.ini, and a `pytest.importorskip` tot he test file, for those that don't have libarchive.
- this repo enforces style, so please run `black` and `flake8` to get the CI to pass.

**Comment by martindurant** [comment]
ping @juliospain - are you still interested in finishing this PR?

**Comment by juliospain** [comment]
You can close it for now if you want. I still want to give it a go, but I need to find some time first.

**Comment by martindurant** [comment]
Happy to wait, if you still think you might get to it

**Comment by juliospain** [comment]
Ok, now it should be ready.

The followings formats should be working (but onky read mode) from the libarchive webpage:

tar (read and write, including GNU extensions)
pax (read and write, including GNU and star extensions)
cpio (read and write, including odc, newc, and afio variants)
ISO9660 (read and write, including Joliet and Rockridge extensions, with some limitations)
zip (read and write, with some limitations, uses zlib)
mtree (read and write, uses separate libraries for creating and verifying cryptographic hashes)
shar (write only)
ar (read and write, including BSD and GNU/SysV variants)
empty (read only; in particular, note that no other format will accept an empty file)
raw (read only)
xar (read and write)
lha/lzh (read only)
rar (read only, original and RAR v5 format)
Microsoft CAB format (read only)
7-Zip (read and write)
WARC (ISO 28500:2009, read and write

File: fsspec/implementations/tests/test_libarchive.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        print(fs.dir_cache)
+
+        assert list(m) == ["a", "b", "deeply/nested/path"]
+        assert m["b"] == data["b"]
+
+
+def test_pickle():
+    with temparchive(data) as archive_file:
+        fs = fsspec.filesystem("libarchive", fo=archive_file)
+        fs2 = pickle.loads(pickle.dumps(fs))
```
Comment (martindurant) [inline fsspec/implementations/tests/test_libarchive.py:55]: I wonder, `fs2 is fs`? Instances are cached by `fsspec`, it may well be.

**Review by martindurant — COMMENTED** [review]
Please add libarchive to the API docs under builtin implementations!

I would put a link to the list of supported archive formats in the class docstring.

All looks good.

**Pull request closed without merging**